logging: enable setting custom logging output func

ldenefle opened the pull request:
To enable custom formatting of the log output while still using existing
backend, this commit adds the `log_output_custom` feature.
To use, register a commit with `log_custom_output_msg_set` then set the
log backend format set using `log_backend_format_set` with
`LOG_OUTPUT_CUSTOM`

ldenefle committed:
logging: enable setting custom logging output func

To enable custom formatting of the log output while still using existing
backend, this commit adds the `log_output_custom` feature.
To use, register a commit with `log_custom_output_msg_set` then set the
log backend format set using `log_backend_format_set` with
`LOG_OUTPUT_CUSTOM`

Signed-off-by: Lucas Denefle <lucas.denefle@converge.io>

nordic-krch commented on tests/subsys/logging/log_custom_format/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
imo it's better to extend `log_switch_format` suite than adding new one.
Note that in `testcase.yaml` of that test you can setup a configuration with and without custom formatting and then skip the test based on presence of `CONFIG_LOG_CUSTOM_SUPPORT`)

ldenefle replied:
I started going down that route indeed then switch to a its own testcase. Will do that then

nordic-krch commented on subsys/logging/Kconfig.formatting:
```diff
@@ -109,6 +109,14 @@ config LOG_DICTIONARY_SUPPORT
 	  This should be selected by the backend automatically.
 
+config LOG_CUSTOM_SUPPORT
+	bool "Custom support"
```
too generic name. Something like `LOG_CUSTOM_FORMAT_SUPPORT` and `Custom formatter support` would be better.

nordic-krch commented on .gitignore:
```diff
@@ -62,3 +62,5 @@ TAGS
 
 .idea
+
+.ccls-cache
```
please move that to another PR. It's not related to the logging changes.

ldenefle commented on include/zephyr/logging/log_output_custom.h:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
```
I'm not too sure what my copyright is ? I work for a company called Converge.
Should I just stick that name there ?

nordic-krch replied:
yes, then you should add this copyright

ldenefle commented:
I haven't completely figured out how you guys like to review PR, for now I added the modifications requested by either you or the compliance scripts as a big fixup that I intend to squash as part of the rebase later. 
I reverted the .ccls-cache commit. I'll make sure that the squashing remove it.

nordic-krch commented:
Usually we append changes to commits instead of adding new commits which get squashed.
Looks good, please fix CI failures.

ldenefle commented:
I did the requested modifications, not sure why the CI was not passing it has something to do with the `esp32` `likely` macro but don't understand why my changes would trigger a build error. I can't reproduce locally with twister when running against the faulty platforms.
I rebased on latest upstream to see if it fixes it.

ldenefle commented:
I think the issue was the include order in the tests, which was set by the linter I pushed something, we'll see

ldenefle commented:
Hello, can I do something to help speed up the merge of that PR ?
